server/db/FirebaseDB.go

```diff
@@ -0,0 +1,101 @@
+package db
+
+import (
+	"errors"
+
+	"github.com/google/uuid"
+	"github.com/jak103/uno/model"
+)
+
+type firebaseDB struct {
+	games         map[string]model.Game
+	gamePasswords map[string]model.Game
+	players       map[string]model.Player
+}
+
+// HasGame checks to see if a game with the given ID exists in the database.
+func (db *firebaseDB) HasGameByPassword(password string) bool {
+	_, ok := db.gamePasswords[password]
+	return ok
+}
+
+// HasGameByID checks to see if a game with the given ID exists in the database.
+func (db *firebaseDB) HasGameByID(id string) bool {
+	_, ok := db.games[id]
+	return ok
+}
+
+// CreateGame a game with the given ID. Perhaps this should instead just return an id?
+func (db *firebaseDB) CreateGame() (*model.Game, error) {
+	myGame := model.Game{ID: uuid.New().String(), Password: "12234"}
+	db.games[myGame.ID] = myGame
+	db.gamePasswords[myGame.Password] = myGame
+	return &myGame, nil
+}
+
+// CreatePlayer creates the player in the database
+func (db *firebaseDB) CreatePlayer(name string) (*model.Player, error) {
+	player := model.Player{ID: uuid.New().String(), Name: name}
+	db.players[player.ID] = player
+	return &player, nil
+}
+
+// LookupGameByID looks up an existing game in the database.
+func (db *firebaseDB) LookupGameByID(id string) (*model.Game, error) {
+	if game, ok := db.games[id]; ok {
+		return &game, nil
+	}
+	return nil, errors.New("mockdb: game not found")
+}
+
+// LookupGameByPassword looks up an existing game in the database.
+func (db *firebaseDB) LookupGameByPassword(password string) (*model.Game, error) {
+	if game, ok := db.gamePasswords[password]; ok {
+		return &game, nil
+	}
+	return nil, errors.New("mockdb: game not found")
+}
+
+// LookupPlayer checks to see if a player is in the database
+func (db *firebaseDB) LookupPlayer(id string) (*model.Player, error) {
+	if player, ok := db.players[id]; ok {
+		return &player, nil
+	}
+	return nil, errors.New("mockdb: player not found")
+}
```

Comment:
Ryan. Just to confirm your question. Are you asking about `LookupPlayer` being declared multiple times in the different files?

Reply:
This just checking for a value in a map, and I understand the hesitation with it being repeated so much. We could write a function that looks for a key and returns `nil` if passed a map, but we still have the added complication of returning different types from each function and map. It really depends on where we want to spend our time.

Reply:
@Ryan-Johnson-1315 . To explain what I think your question is.  I apologize that this is going to wax a little verbose, but I just thought to error on the side of extra information than less information. If you happen to know all of this previously, I apologize.

What I imagine your question is, why are there so many many files have been created in the DB section. And why do some of these files appear to be mostly duplicates of each other.

So, in a lot of Object Oriented Languages, like Java, C++, C#, you have the option for method overloading on an abstract class.  In Go, there isn't the idea of a base class, or an abstract class. Basically we just define an interface, and then every struct that we establish, we need to entirely implement each method in order to honor the interface.

What this means is that you can't have a `default` way of handling things persay. you basically have to define each method individually (even if they are duplicates of each other)

However, the case of what we are trying to build now is somewhat uncommon.  We've established a single interface for all of our database interactions, and a series of subclasses that handle `different` types of database interactions. (Mongo, Firestore, and Mock/Unit tests).  

Typically  in most applications you maintain the same database type on dev as well as on prod. However, in our case we need to support `similar` database types, but not identical. Which means we need to translate application logic into database logic for each type of database. (So for example, normally you would just have one file to handle Database transactions, but if you are essentially supporting 3 different databases, you have to write 3 files, and establish a contract/interface that they are all required to honor) 

However... in our case, @hhenrichsen and I did not have all of the tools/items on hand to build out the specific database logic for interacting with Firestore. So we opted to currently have FirestoreDB be a duplicate of what our MockDB is accomplishing.  The MockDB is going to only be used for testing purposes, but it serves as a useful `stub` for our Firestore Connections until we finish those out.

Eventually after we get this shipped, we will work with @jak103 to actually round out the functionality for the FirestoreDB struct to be specific to Firestore itself.

Reply:
Sounds good. I have been looking at a few implementations of how to do it it would behave as a overridden function, but I believe you are correct. I think this is the best way to do it. Everything else looks great!